### PR TITLE
[twitcasting] Refactor extractor

### LINKED ISSUE #382: [TwitCasting] Some archive downloads only the first hour of it
## Checklist

- [ ] I'm reporting a broken site support
- [x] I've verified that I'm running yt-dlp version **2021.06.01**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
WIP
```

## Description/Background
For some TwitCasting archive, it sometimes split the archive 1 hour each.
This especially happens when the video is longer than an hour, with high bitrate.

Here's an example video: https://twitcasting.tv/loft_heaven/movie/685979292
And another one: https://twitcasting.tv/nana_fujisawa/movie/686749195 

With current yt-dlp, it can only download the first split(=hour). But there's more big problem:
1. TC allows downloading only one stream simultaneously. Failing to follow this **causes all playlists to expire.**
    - Some `ffmpeg` hack like `ffmpeg -i STREAM1.m3u8 -i STREAM2.m3u8 -filter_complex "blablabla" -map [v] -map [a] outpur.mp4.part` don't work because of this.
    - It seems we have to request the next playlist after we download last segment of previous playlist.
    - (More to say, it should be realtime. Download breaks sometimes)
1. ~~Current youtube-dl/yt-dlp architecture doesn't support such videos.~~ ... but still not able to concat them.
    - I think it's okay to ask users to download these splits manually, but who likes it?
    - They do not have a proper method to download multiple videos separately and combine later.

I'm also trying to work better with these problems, but I want to hear idea from you.

Thanks.

### PR DESCRIPTION
**Pull request by Lesmiscore**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Table of Contents
- Find multiple m3u8 playlists to download all parts of archive (https://github.com/yt-dlp/yt-dlp/issues/382)
- Add thumbnail extraction
- Fix duration extraction
- Extract source format if available
- Closes #382 (must be used with `--concat-playlist` to satisfy your needs)

### CODE REVIEW

File: yt_dlp/extractor/twitcasting.py [hunk 1]
```diff
@@ -79,61 +88,116 @@ def _real_extract(self, url):
...
-                webpage, 'movie playlist', group='url', default='[{}]'), video_id)
-            if isinstance(video_js_data, dict):
-                video_js_data = list(video_js_data.values())[0]
-            video_js_data = video_js_data[0]
-            m3u8_url = try_get(video_js_data, lambda x: x['source']['url'])
+        video_js_data = try_get(
+            webpage,
+            lambda x: self._parse_json(self._search_regex(
+                r"data-movie-playlist='([^']+?)'",
+                x, 'movie playlist', default=None), video_id)["2"], list)
```
Comment (Ashish0804) [inline yt_dlp/extractor/twitcasting.py:95]:
```suggestion
        video_js_data = try_get(
            webpage,
            lambda x: self._parse_json(self._search_regex(
                r'data-movie-playlist=\'([^\']+?)\'',
                x, 'movie playlist', default=None), video_id)['2'], list)
```
I prefer using `'` and just escaping the other single quotes present in the pattern.

File: yt_dlp/extractor/twitcasting.py [hunk 2]
```diff
@@ -7,19 +7,19 @@
 from .common import InfoExtractor
 from ..downloader.websocket import has_websockets
 from ..utils import (
+    ExtractorError,
     clean_html,
```
Comment (Ashish0804) [inline yt_dlp/extractor/twitcasting.py:11]: did u mean this?
```suggestion
    clean_html,
    ExtractorError,
```

Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py:11]: yes

File: yt_dlp/extractor/twitcasting.py (hunk 1 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py:95]: yeah it should

File: yt_dlp/extractor/twitcasting.py [hunk 3]
```diff
@@ -79,61 +88,116 @@ def _real_extract(self, url):
...
+            list)
+        if not m3u8_urls:
+            raise ExtractorError('Failed to get m3u8 playlist')
+
+        if is_live:
+            m3u8_url = m3u8_urls[0]
+            formats = self._extract_m3u8_formats(
+                m3u8_url, video_id, ext='mp4', m3u8_id='hls',
+                live=True, headers={
+                    'Accept': '*/*',
```
Comment (Ashish0804) [inline yt_dlp/extractor/twitcasting.py:127]: isn't this present by default? 🤔

File: yt_dlp/extractor/twitcasting.py [hunk 4]
```diff
@@ -79,61 +88,116 @@ def _real_extract(self, url):
...
+                    'url': m3u8_url,
+                    'ext': 'mp4',
+                    # has to be ffmpeg because the requesting them at once will cause download failure
+                    # (https://github.com/yt-dlp/yt-dlp/issues/382)
+                    'protocol': 'm3u8',
+                    'http_headers': {
+                        'Accept': '*/*',
+                        'Origin': 'https://twitcasting.tv',
+                        'Referer': 'https://twitcasting.tv/',
+                    },
```
Comment (Ashish0804) [inline yt_dlp/extractor/twitcasting.py:177]: u are using these again and again.
Just set a var as header and pass that everywhere

File: yt_dlp/extractor/twitcasting.py (hunk 3 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py:127]: I don't think so

**Review by Ashish0804 — APPROVED** [review]

File: yt_dlp/extractor/twitcasting.py (hunk 3 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py]: actually not needed
removed

File: yt_dlp/extractor/twitcasting.py [hunk 5]
```diff
@@ -79,74 +93,122 @@ def _real_extract(self, url):
...
-                    'quality': qq(mode),
-                    'protocol': 'websocket_frag',  # TwitCasting simply sends moof atom directly over WS
-                })
-
-        thumbnail = video_js_data.get('thumbnailUrl') or self._og_search_thumbnail(webpage)
+        thumbnail = try_get(
+            video_js_data,
+            (lambda x: traverse_obj(x, (0, 'thumbnailUrl')),
+             lambda x: self._og_search_thumbnail(webpage)),
+            str)
```
Comment (pukkandan) [inline yt_dlp/extractor/twitcasting.py:106]:
```suggestion
        thumbnail = traverse_obj(video_js_data, (0, 'thumbnailUrl')) or self._og_search_thumbnail(webpage)
```
why try_get?

File: yt_dlp/extractor/twitcasting.py [hunk 6]
```diff
@@ -79,74 +93,122 @@ def _real_extract(self, url):
...
             ['description', 'og:description', 'twitter:description'], webpage)
-        duration = float_or_none(video_js_data.get(
-            'duration'), 1000) or parse_duration(clean_html(
-                get_element_by_class('tw-player-duration-time', webpage)))
+        duration = try_get(
+            video_js_data,
+            (lambda x: sum(float_or_none(y.get('duration')) for y in x) / 1000,
+             lambda x: parse_duration(clean_html(
+                 get_element_by_class('tw-player-duration-time', webpage)))),
+            float)
```
Comment (pukkandan) [inline yt_dlp/extractor/twitcasting.py:115]:
```suggestion
        duration = (try_get(video_js_data, lambda x: sum(float_or_none(y.get('duration')) for y in x) / 1000)
                    or parse_duration(clean_html(get_element_by_class('tw-player-duration-time', webpage))))
```

File: yt_dlp/extractor/twitcasting.py [hunk 7]
```diff
@@ -79,74 +93,122 @@ def _real_extract(self, url):
...
+            if data_movie_url:
+                return [data_movie_url]
+
+        m3u8_urls = try_get(
+            webpage,
+            (find_dmu,
+             lambda x: traverse_obj(video_js_data, (..., 'source', 'url')),
+             lambda x: ['https://twitcasting.tv/%s/metastream.m3u8' % uploader_id]
+                if is_live else None),
+            list)
```
Comment (pukkandan) [inline yt_dlp/extractor/twitcasting.py:154]: again, why wrap everything in try_get unnecessarily?
```suggestion
        m3u8_urls = (try_get(webpage, find_dmu, list)
                     or traverse_obj(video_js_data, (..., 'source', 'url'))
                     or [f'https://twitcasting.tv/{uploader}/metastream.m3u8'] if is_live else None)
```

File: yt_dlp/extractor/twitcasting.py [hunk 8]
```diff
@@ -79,74 +93,122 @@ def _real_extract(self, url):
...
+                'formats': formats
+            }
+        else:
+            infodict = {
+                '_type': 'multi_video',
+                'entries': [{
+                    'id': f'{video_id}-{num}',
+                    'url': m3u8_url,
+                    'ext': 'mp4',
+                    # has to be ffmpeg because the requesting them at once will cause download failure
```
Comment (pukkandan) [inline yt_dlp/extractor/twitcasting.py:198]: wdym? native downloader doesn't request them all at once

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/twitcasting.py (hunk 5 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py:106]: it used to be just indexing where `try_get` is applicable

File: yt_dlp/extractor/twitcasting.py (hunk 8 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py:198]: actually these manifests require `_extract_m3u8_formats`, but using it in extractor results in `the requesting them at once will cause download failure`

File: yt_dlp/extractor/twitcasting.py [hunk 9]
```diff
@@ -79,74 +93,110 @@ def _real_extract(self, url):
...
+                live=True, headers=self._M3U8_HEADERS)
+
+            try:
+                formats.extend(self._extract_m3u8_formats(
+                    m3u8_url, video_id, ext='mp4', m3u8_id='source',
+                    live=True, query={'mode': 'source'},
+                    note='Downloading source quality m3u8',
+                    headers=self._M3U8_HEADERS))
+            except ExtractorError as ex:
+                self.report_warning(ex)
```
Comment (pukkandan) [inline yt_dlp/extractor/twitcasting.py:159]: just use `fatal=False`?

Comment (Lesmiscore) [reply yt_dlp/extractor/twitcasting.py]: you're right
blaming myself at the time I wrote that

File: yt_dlp/extractor/twitcasting.py [hunk 10]
```diff
@@ -79,74 +93,110 @@ def _real_extract(self, url):
...
+            }
+        else:
+            infodict = {
+                '_type': 'multi_video',
+                'entries': [{
+                    'id': f'{video_id}-{num}',
+                    'url': m3u8_url,
+                    'ext': 'mp4',
+                    # has to be ffmpeg because the requesting them at once will cause download failure
+                    # (https://github.com/yt-dlp/yt-dlp/issues/382)
```
Comment (pukkandan) [inline yt_dlp/extractor/twitcasting.py:184]:
```suggestion
                    # Requesting the manifests here will cause download to fail.
                    # So use ffmpeg instead. See: https://github.com/yt-dlp/yt-dlp/issues/382
```

**Pull request merged**